pkg/tracking/store/sql/experiments_search_builder.go

```diff
@@ -0,0 +1,278 @@
+	var offset int
+
+	if pageToken != "" {
+		var token PageToken
+		if err := json.NewDecoder(
+			base64.NewDecoder(
+				base64.StdEncoding,
+				strings.NewReader(pageToken),
+			),
+		).Decode(&token); err != nil {
+			return nil, 0, contract.NewError(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf("invalid page_token '%s': %s", pageToken, err),
+			)
+		}
+
+		offset = int(token.Offset)
+	}
+
+	return query.Offset(offset), offset, nil
+}
+
+func applyExperimentsLifecycleStagesFilter(query *gorm.DB, runViewType protos.ViewType) *gorm.DB {
+	switch runViewType {
+	case protos.ViewType_ACTIVE_ONLY:
+		query = query.Where(
+			"lifecycle_stage IN (?)", []models.LifecycleStage{
+				models.LifecycleStageActive,
+			},
+		)
+	case protos.ViewType_DELETED_ONLY:
+		query = query.Where(
+			"lifecycle_stage IN (?)", []models.LifecycleStage{
+				models.LifecycleStageDeleted,
+			},
+		)
+	case protos.ViewType_ALL:
+		query = query.Where(
+			"lifecycle_stage IN (?)", []models.LifecycleStage{
+				models.LifecycleStageActive,
+				models.LifecycleStageDeleted,
+			},
+		)
+	}
+
+	return query
+}
+
+func applyExperimentsOrderBy(query *gorm.DB, orderBy []string) (*gorm.DB, *contract.Error) {
+	expOrder := false
+
+	for _, o := range orderBy {
+		parts := experimentOrder.FindStringSubmatch(o)
+		if len(parts) == 0 {
+			return nil, contract.NewError(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf("invalid order_by clause '%s'", o),
+			)
+		}
+
+		column := parts[1]
+		switch column {
+		case "experiment_id":
+			expOrder = true
+
+			fallthrough
+		case "name", "creation_time", "last_update_time":
+		default:
+			return nil, contract.NewError(
+				protos.ErrorCode_INVALID_PARAMETER_VALUE,
+				fmt.Sprintf(
+					"invalid attribute '%s'. Valid values are ['name', 'experiment_id', 'creation_time', 'last_update_time']",
+					column,
+				),
+			)
+		}
+
+		query = query.Order(clause.OrderByColumn{
+			Column: clause.Column{Name: column},
+			Desc:   len(parts) == 3 && strings.ToUpper(parts[2]) == "DESC",
+		})
+	}
+
+	if len(orderBy) == 0 {
+		query = query.Order("experiments.creation_time DESC")
+	}
+
+	if !expOrder {
+		query = query.Order("experiments.experiment_id ASC")
+	}
+
+	return query, nil
+}
+
+//nolint:funlen,gocognit,nestif,cyclop,goconst,mnd,forcetypeassert
+func applyExperimentsFilter(database, query *gorm.DB, filter string) (*gorm.DB, *contract.Error) {
+	if filter != "" {
+		for index, f := range filterAnd.Split(filter, -1) {
+			parts := filterCond.FindStringSubmatch(f)
+			if len(parts) != 5 {
+				return nil, contract.NewError(
+					protos.ErrorCode_INVALID_PARAMETER_VALUE,
+					fmt.Sprintf("malformed filter '%s'", f),
+				)
+			}
+
+			var value any = parts[4]
+
+			entity, key, comparison := parts[1], strings.Trim(parts[2], "\"`"), parts[3]
+
+			switch entity {
+			case "", "attribute", "attributes", "attr":
```

Comment:
I have some of these strings as constants in `/pkg/tracking/service/query/parser/validate.go`.
Perhaps we want a more generate mlflow constants package at some point?
Could be addressed in the future though.